helm-chart/templates/rules.yaml

```diff
@@ -0,0 +1,22 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: {{ include "ttjs.fullname" . }}
+  namespace: {{ quote (or .Values.serviceMonitor.namespace .Release.Namespace) }}
+  labels:
+    {{- include "ttjs.labels" . | nindent 4 }}
+    {{- range $key, $value := .Values.rules.labels }}
+    {{ $key }}: {{ $value | quote }}
+    {{- end }}
+spec:
+  groups:
+    - name: ttjs.rules
+      rules:
+        - alert: TheThingsJoinServerNoJoinsAccepted
+          expr: sum by (namespace, job) (rate(ttjs_activations_total{error_type=""}[10m])) == 0
+          for: 15m
```

Comment:
Users deploying this thing without devices, or a few devices that hardly ever join, will have this alert firing all the time. We can make this configurable, but then we change the rule definition for a release, and then we need to segment rules per namespace and release, see "notes for reviewers".

Reply:
Is `job` here the pod ID, or the replica set ID ? (it is the task ID in ECS)

I think that's too granular if we look at the rates per pod.

---

Regarding the expression itself - in TTS we have some places in which we compare the long running average for a day (+2 standard deviations, i.e. 95% of the observed rate values) with the average over the past 10 minutes. This is somewhat smart because it detects if values observed in the last 5 minutes are in the other 5% of the values.

You can see an example of how such an expression would look like [here](https://github.com/TheThingsIndustries/lorawan-stack-aws/blob/e249ee81b9868899a8ee73a5663b39f0ec235233/prometheus/rules/application-server/application-server.yml#L100-L107). It requires some extra recording rules for the daily average and standard deviation, and also for the shorter term average, but it is a bit more robust.

Reply:
The `job` is the deployment name, which is the `fullname` template helper from Helm, which is essentially the release name. With Helm releases, we need two identifiers to uniquely identify them: the namespace (like `joinserver`) and the release name (like `ttjs`).

---

I don't see how that solves the issue of firing alerts unnecessarily because there are too few join-requests in general. The question is how do we generalize an expression to fire an alert when there are 0 joins accepted and that alert fires fast enough versus there are very few joins accepted because this Join Server is hardly used.

In a busy Join Server where a issue occurs so that joins are no longer accepted, this can be observed immediately. In an idle Join Server, we may need to increase the `for` part to 30 minutes or so. But are those 30 minutes enough for a busy Join Server operator? Or do we need two events with different priorities, one for no joins accepted in 10 minutes and one for no joins accepted in 60 minutes?

Regarding comparing current values with past averages; I wonder if a Join Server is the right place to do that, because it's oftentimes the network that causes rejoins, for instance because of link loss at scale. That is relevant for a network operator but not for a Join Server operator.

Reply:
A watchdog is nice but is quite some work to do right; we need to bootstrap a provisioner that provisions a device, we need a TLS trust store for an AS, we need signed client certs with which we claim the device and send a join-request, ideally with incrementing DevNonce so there's state, and all the responses need to be verified. Very nice, but not in scope here.

For now, I made the alert configurable.